File: code/scala/src/main/scala/group/pals/desktop/app/apksigner/utils/Network.scala
/*
 *    Copyright (C) 2012 Hai Bison
 *
 *    See the file LICENSE at the root directory of this project for copying
 *    permission.
 */

package group.pals.desktop.app.apksigner.utils

import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

import android.util.Base64

/**
 * Network utilities.
 *
 * @author Hai Bison
 * @since v1.6 beta
 */
object Network {

    /**
     * The network timeout, in milliseconds.
     */
    lazy final val NETWORK_TIMEOUT = 15000

    /**
     * HTTP status code OK.
     */
    lazy final val HTTP_STATUS_OK = 200

    /**
     * The redirection status code {@code 302}.
     */
    lazy final val HTTP_STATUS_FOUND = 302

    /**
     * Max redirection allowed.
     */
    lazy final val MAX_REDIRECTION_ALLOWED = 9

    /**
     * Header field "Location".
     */
    lazy final val HEADER_LOCATION = "Location"

    /**
     * The date format of header fields.
     */
    lazy final val HEADER_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss z"

    /**
     * Header field "Expires".
     */
    lazy final val HEADER_EXPIRES = "Expires"

    private lazy final val PROPERTY_SYS_HTTP_PROXY_HOST = "http.proxyHost"
    private lazy final val PROPERTY_SYS_HTTP_PROXY_PORT = "http.proxyPort"

    private lazy final val PROPERTY_SYS_HTTPS_PROXY_HOST = "https.proxyHost"
    private lazy final val PROPERTY_SYS_HTTPS_PROXY_PORT = "https.proxyPort"

    /**
     * Opens new connection to {@code url} with default settings.
     *
     * @param url
     *            the URL.
     * @return the connection. Or {@code null} if an error occurred.
     */
    def openHttpConnection(url: String): HttpURLConnection = {
        if (Texts.isEmpty(url))
            return null

        if (Preferences.usingProxy) {
            /*
             * HTTP
             */
            System.setProperty(PROPERTY_SYS_HTTP_PROXY_HOST,
                Preferences.proxyHost)
            System.setProperty(PROPERTY_SYS_HTTP_PROXY_PORT,
                Preferences.proxyPort.toString())

            /*
             * HTTPS
             */
            System.setProperty(PROPERTY_SYS_HTTPS_PROXY_HOST,
                Preferences.proxyHost)
            System.setProperty(PROPERTY_SYS_HTTPS_PROXY_PORT,
                Preferences.proxyPort.toString())
        } else {
            for (
                s <- Array(PROPERTY_SYS_HTTP_PROXY_HOST,
                    PROPERTY_SYS_HTTP_PROXY_PORT,
                    PROPERTY_SYS_HTTPS_PROXY_HOST,
                    PROPERTY_SYS_HTTPS_PROXY_PORT)
            ) System.clearProperty(s)
        }

        try {
            var conn = new URL(url).openConnection().asInstanceOf[HttpURLConnection]

            if (Preferences.usingProxy
                && !Texts.isEmpty(Preferences.proxyUsername)) {
                var password = Preferences.proxyPassword
                var proxyAuthorization =
                    Base64.encodeToString(
                        (Preferences.proxyUsername + ":"
                            + (if (password != null) new String(password) else ""))
                            .getBytes(Texts.UTF8),
                        Base64.NO_WRAP)
                // https://en.wikipedia.org/wiki/List_of_HTTP_header_fields
                conn.setRequestProperty("Proxy-Authorization", "Basic "
                    + proxyAuthorization)
            }

            conn.setConnectTimeout(NETWORK_TIMEOUT)
            conn.setReadTimeout(NETWORK_TIMEOUT)
            conn
        } catch {
            case e: MalformedURLException =>
                e.printStackTrace()
                null
            case e: IOException =>
                /*
                * Perhaps there is no available Internet connections.
                */
                e.printStackTrace()
                null
            case e: Exception =>
                e.printStackTrace()
                null
        }
    } // openHttpConnection()

}

File: code/jython/apksigner/utils/network.py
#  Copyright (C) 2012 Hai Bison
#
#  See the file LICENSE at the root directory of this project for copying
#  permission.

'''
Network utilities.
'''

import urllib2

from apksigner.utils import settings

'''
The network timeout, in seconds.
'''
NETWORK_TIMEOUT = 15

'''
Max redirection allowed.
'''
MAX_REDIRECTION_ALLOWED = 9

'''
Header field "Location".
'''
HEADER_LOCATION = 'Location'

'''
The date format of header fields.
'''
HEADER_DATE_FORMAT = 'EEE, dd MMM yyyy HH:mm:ss z'

'''
Header field 'Expires'.
'''
HEADER_EXPIRES = 'Expires'

def open_url(url):
    ''' Opens new connection to ``url`` with default settings.

        **NOTE:** Currently this function doesn't work because of issue #2103:
        http://bugs.jython.org/issue2103 -- You should use the Java wrapper
        instead: ``open_java_url(url)``.

        Parameters:

        :url (String):
            the URL.

        Returns:
            the connection, or ``None`` if an error occurred.
    '''

    if not url: return

    if settings.is_using_proxy():
        _ = '{}:{}'.format(settings.get_proxy_host(), settings.get_proxy_port())
        proxy_handler = urllib2.ProxyHandler({ 'http': _, 'https': _ })

        proxy_auth_handler = None
        if settings.get_proxy_username() and settings.get_proxy_password():
            pwd_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
            pwd_mgr.add_password(None, settings.get_proxy_host(),
                                 settings.get_proxy_username(),
                                 settings.get_proxy_password())
            proxy_auth_handler = urllib2.ProxyBasicAuthHandler(pwd_mgr)

        opener = urllib2.build_opener(proxy_handler, proxy_auth_handler)
        try:
            return opener.open(url, timeout=NETWORK_TIMEOUT)
        except:
            # Perhaps there is no available Internet connections.
            pass
        #.if

    try:
        return urllib2.urlopen(url, timeout=NETWORK_TIMEOUT)
    except:
        # Perhaps there is no available Internet connections.
        pass

    #.open_url()

def open_java_url(url):
    ''' Opens new connection to ``url`` with default settings.

        Parameters:

        :url (String):
            the URL.

        Returns:
            the ``java.net.HttpURLConnection``, or ``None`` if an error
            occurred.
    '''

    PROPERTY_SYS_HTTP_PROXY_HOST = 'http.proxyHost'
    PROPERTY_SYS_HTTP_PROXY_PORT = 'http.proxyPort'

    PROPERTY_SYS_HTTPS_PROXY_HOST = 'https.proxyHost'
    PROPERTY_SYS_HTTPS_PROXY_PORT = 'https.proxyPort'

    if not url: return

    # Proxies

    from java.lang import System
    if settings.is_using_proxy():
        System.setProperty(PROPERTY_SYS_HTTP_PROXY_HOST,
                           settings.get_proxy_host())
        System.setProperty(PROPERTY_SYS_HTTP_PROXY_PORT,
                           str(settings.get_proxy_port()))

        System.setProperty(PROPERTY_SYS_HTTPS_PROXY_HOST,
                           settings.get_proxy_host())
        System.setProperty(PROPERTY_SYS_HTTPS_PROXY_PORT,
                           str(settings.get_proxy_port()))
    else:
        for s in [ PROPERTY_SYS_HTTP_PROXY_HOST, PROPERTY_SYS_HTTP_PROXY_PORT,
                   PROPERTY_SYS_HTTPS_PROXY_HOST, PROPERTY_SYS_HTTPS_PROXY_PORT ]:
            System.clearProperty(s)

    # Now create connection

    try:
        from java.net import URL
        conn = URL(url).openConnection()

        if settings.is_using_proxy():
            if settings.get_proxy_username() and settings.get_proxy_password():
                import base64
                proxy_auth = base64.urlsafe_b64encode('{}:{}'.format(
                    settings.get_proxy_username(),
                    settings.get_proxy_password()))
                # https://en.wikipedia.org/wiki/List_of_HTTP_header_fields
                conn.setRequestProperty('Proxy-Authorization',
                                        'Basic ' + proxy_auth)

        conn.setConnectTimeout(NETWORK_TIMEOUT * 1000)
        conn.setReadTimeout(NETWORK_TIMEOUT * 1000)
        return conn
    except:
        # Perhaps there is no available Internet connections.
        pass
    #.open_java_url()